ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
       numChunksRetrieved.set(0);
     }
   }
+
+  private class CachedFirstChunk extends FirstGetChunk {
+
+    /**
+     * Construct a CachedFirstChunk and initialize it with the {@link BlobMetadata} of the overall operation.
+     * @param blobMetadata Cached blob metadata for composite blobs
+     */
+    CachedFirstChunk(BlobMetadata blobMetadata) {
+      super();
+      /**
+       * If cache hit, extract all relevant metadata and save in appropriate variables.
+       * If the entry gets evicted later, we are still safe as we have saved all the metadata.
+       */
+      blobType = BlobType.MetadataBlob;
+      compositeBlobInfo = blobMetadata.getCompositeBlobInfo();
+      totalSize = compositeBlobInfo.getTotalSize();
+      chunkMetadataList = compositeBlobInfo.getChunkMetadataList();
+      blobInfo = blobMetadata.getBlobInfo();
+      lifeVersion = blobInfo.getLifeVersion();
+      serverBlobProperties = blobInfo.getBlobProperties();
+      chunkIndexToBuf = new ConcurrentHashMap<>();
+      chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+      successfullyDeserialized = true;
+      logger.info(String.format("Metadata cache-hit for blobId = %s", blobId));
+    }
+
+    /**
+     * Return the {@link MessageFormatFlags} to associate with the first getBlob chunk operation.
+     * @return {@link MessageFormatFlags#Blob} for {@link GetBlobOptions.OperationType#Data}, or
+     *         {@link MessageFormatFlags#All} by default.
+     */
+    @Override
+    MessageFormatFlags getOperationFlag() {
+      /**
+       * If we had a metadata cache hit, then we just need to get blobInfo to validate the cache entry.
+       * We don't have to fetch the entire metadata again.
+       */
+      return MessageFormatFlags.BlobInfo;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br>
+     * It would help to keep in mind while going through this method that the first chunk is either a metadata chunk
+     * or the only chunk of the blob.
+     */
+    @Override
+    void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
+        throws IOException, MessageFormatException {
+
+      /**
+       * If we find blob metadata in frontend cache, validate it and use it.
+       */
+      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      if (!serverBlobProperties.toString().equals(receivedBlobProperties.toString())) {
```

Comment:
verify the equality of blob properties should be done in a different way. we should implement an `equal` method for BlobProperties.

Reply:
got it. I thought of the same first !